### PR TITLE
fix: resolve migration issues with Spanner

### LINKED ISSUE #9725: Migrations do not work with Spanner
### Issue description

Because Spanner does not offer generated id columns, inserts into the `migrations` table fail because the `id` column is null.

### Expected Behavior

When using the Spanner driver, a valid Migration should be applied and a record should be inserted into the `migrations` table.

### Actual Behavior

The migration itself runs fine (in my case a table is created), but the insert into `migrations` table fails due to a null `id` field.

### Steps to reproduce

1. Set up a Spanner DataSource:
```typescript
import { DataSource } from "typeorm";
import { User } from "entities/user";

export const AppDataSource = new DataSource({
  type: "spanner",
  projectId: process.env.PROJECT_ID,
  instanceId: process.env.SPANNER_INSTANCE_ID,
  databaseId: process.env.SPANNER_DATABASE_ID,
  logging: true,
  entities: [
    User,
  ],
  subscribers: [],
  migrations: [
    initialTableCreation1674233117160
  ],
  migrationsRun: true,
  migrationsTableName: "migrations"
})
```
2. Create a migration:
```typescript
import { MigrationInterface, QueryRunner, Table } from "typeorm";

export class initialTableCreation1674233117160 implements MigrationInterface {
  name = 'initialTableCreation1674233117160'

  public async up(queryRunner: QueryRunner): Promise<void> {
    await queryRunner.createTable(new Table({
      name: "User",
      columns: [
        {
          name: "shardId",
          type: "int64",
          isPrimary: true
        },
        {
          name: "name",
          type: "string"
        },
        {
          name: "email",
          type: "string"
        }
      ]
    }), true);
  }

  public async down(queryRunner: QueryRunner): Promise<void> {
    await queryRunner.dropTable(`User`, true);
  }
}
```
3. Ensure the following line (or equivalent) is in your app's `index.ts` or start up file:
```typescript
(async () => {
  await AppDataSource.initialize();
})();
```
4. Also ensure, if you are running Spanner in an emulator the `SPANNER_EMULATOR_HOST` variable is set appropriately.
5. Start your app - The logs should show `query failed:` on the insert into `migrations`:
```
error: Error: 9 FAILED_PRECONDITION: Cannot specify a null value for column: migrations.id in table: migrations
```

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System    |  macOS Monterey 12.6 |
| Node.js version     | 16.16.0  |
| Typescript version  | 4.9.4 |
| TypeORM version     | 0.3.11  |

### Additional Context

_No response_

### Relevant Database Driver(s)

- [ ] aurora-mysql
- [ ] aurora-postgres
- [ ] better-sqlite3
- [ ] cockroachdb
- [ ] cordova
- [ ] expo
- [ ] mongodb
- [ ] mysql
- [ ] nativescript
- [ ] oracle
- [ ] postgres
- [ ] react-native
- [ ] sap
- [X] spanner
- [ ] sqlite
- [ ] sqlite-abstract
- [ ] sqljs
- [ ] sqlserver

### Are you willing to resolve this issue by submitting a Pull Request?

Yes, I have the time, but I don't know how to start. I would need guidance.

### PR DESCRIPTION
**Pull request by ikenox**
### Description of change

As reported by #9725, migration doesn't work with Spanner because it doesn't support auto-increment column.
This pull request fixes the problem by specifying migration id explicitly.

Fixes #9725

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run format` to apply prettier formatting
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [x] There are new or updated unit tests validating the change
- [ ] Documentation has been updated to reflect this change
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: src/migration/MigrationExecutor.ts [hunk 1]
```diff
@@ -647,6 +657,11 @@ export class MigrationExecutor {
         migration: Migration,
     ): Promise<void> {
         const values: ObjectLiteral = {}
+        if (this.connection.driver.options.type === "spanner") {
+            // spanner doesn't support auto-increment column
+            values["id"] = Date.now() + "-" + migration.name
```
Comment (hinogi) [inline src/migration/MigrationExecutor.ts:662]:
```suggestion
            values["id"] = Date.now().toString() + "-" + migration.name
```

**Review by hinogi — APPROVED** [review]
LGTM but I do not know spanner.

**Comment by ikenox** [comment]
@hinogi Thanks, I fixed the code following to your suggestion.

**Comment by ikenox** [comment]
Cloud Spanner now supports auto-increment column.
https://cloud.google.com/spanner/docs/primary-key-default-value#bit-reversed-sequence

I think it's better to use this feature, so I close this pull request.

**Pull request closed without merging**